core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -206,6 +206,58 @@ private[zio] final class FiberContext[E, A](
 
   private[this] final def executor: Executor = executors.peekOrElse(platform.executor)
 
+  @inline private[this] final def raceWithImpl[R, EL, ER, E, A, B, C](
+    race: ZIO.RaceWith[R, EL, ER, E, A, B, C]
+  ): ZIO[R, E, C] = {
+    @inline def complete[E0, E1, A, B](
+      winner: Fiber[E0, A],
+      loser: Fiber[E1, B],
+      cont: (Exit[E0, A], Fiber[E1, B]) => ZIO[R, E, C],
+      winnerExit: Exit[E0, A],
+      ab: AtomicBoolean,
+      cb: ZIO[R, E, C] => Unit
+    ): Unit =
+      if (ab.compareAndSet(true, false)) {
+        winnerExit match {
+          case exit: Exit.Success[_] =>
+            cb(winner.inheritFiberRefs.flatMap(_ => cont(exit, loser)))
+          case exit: Exit.Failure[_] =>
+            cb(cont(exit, loser))
+        }
+      }
+
+    val raceIndicator = new AtomicBoolean(true)
+
+    val left  = fork[EL, A](race.left.interruptible.asInstanceOf[IO[EL, A]])
+    val right = fork[ER, B](race.right.interruptible.asInstanceOf[IO[ER, B]])
+
+    supervise(left)
+    supervise(right)
+
+    ZIO
+      .effectAsync[R, E, C] { cb =>
+        val leftRegister = left.register0 {
+          case exit0: Exit.Success[Exit[EL, A]] =>
+            complete[EL, ER, A, B](left, right, race.leftWins, exit0.value, raceIndicator, cb)
+          case exit: Exit.Failure[_] => complete(left, right, race.leftWins, exit, raceIndicator, cb)
+        }
+
+        if (leftRegister ne null)
+          complete(left, right, race.leftWins, leftRegister, raceIndicator, cb)
+        else {
+          val rightRegister = right.register0 {
+            case exit0: Exit.Success[Exit[_, _]] =>
+              complete(right, left, race.rightWins, exit0.value, raceIndicator, cb)
+            case exit: Exit.Failure[_] => complete(right, left, race.rightWins, exit, raceIndicator, cb)
+          }
+
+          if (rightRegister ne null)
+            complete(right, left, race.rightWins, rightRegister, raceIndicator, cb)
+        }
+      }
+      .onInterrupt(left.interrupt *> right.interrupt)
```

Comment:
This will get faster after the supervision work I'm doing. 

Note you can do a little micro-optimization, replacing `left.interrupt *> right.interrupt` with `left.interrupt.flatMap(_ => right.interrupt)`.